Add `postprocessor::Sequence`

mishig25 opened the pull request:
Closes https://github.com/huggingface/tokenizers/issues/873 by implementing `preprocessor::Sequence`

[as suggested](https://github.com/huggingface/tokenizers/issues/873#issuecomment-1139511282) by @SaulLu, 
> Let me just re-share here a design idea that we had - I think - also discussed offline a long time ago: add a new method to all processors - for example named process_chain - which will have another composable signature (a list) and which will be the method called by processors.Sequence. 😊


Please let me know if proposed changes look good, if so, I will add tests. Otherwise, I'm happy to make any changes as necessary. Also, please see the comments as I had some questions

mishig25 commented on tokenizers/src/tokenizer/mod.rs:
```diff
@@ -100,6 +104,14 @@ pub trait PostProcessor {
         add_special_tokens: bool,
     ) -> Result<Encoding>;
+    /// Process method that is used when a processor is part of `processors::sequence::Sequence`
+    fn process_chain(
+        &self,
+        _encodings: Vec<Encoding>,
+        _add_special_tokens: bool,
+    ) -> Result<Vec<Encoding>> {
```
@Narsil this is the preliminary signature we agreed upon. Please check its implementation & let me know if there is anything you wanna change

mishig25 commented on tokenizers/src/processors/template.rs:
```diff
@@ -647,6 +648,26 @@ impl PostProcessor for TemplateProcessing {
+        encodings: Vec<Encoding>,
+        add_special_tokens: bool,
+    ) -> Result<Vec<Encoding>> {
+        let is_pair = encodings.len() == 2;
+        if is_pair {
```
wdyt about this implementation & hardcoded number 2. 
Basically, I'm saying: if there are 2 elements in `encodings` vector, then run `self.process(encoding, pair_encoding)` with **pair_encoding** encoding.

otherwise, run `self.process(encoding, pair_encoding=None)` for each element in `encodings`

thomasw21 commented:
Not sure I understand why we use `process_chain` ... I think all `fn process` should take `encodings: Vec<Encoding>` instead, and `Sequence` is just a for loop that calls `process` instead of adding a new endpoint?

mishig25 commented on tokenizers/src/processors/sequence.rs:
```diff
@@ -0,0 +1,138 @@
+                .ok_or(SequenceProcessorError::EncodingsVecPop)?),
+            _ => {
+                // merge encodings
+                for (i, encoding) in encodings.iter_mut().enumerate() {
+                    encoding.set_sequence_id(i);
+                }
+                Ok(Encoding::merge(encodings, false))
```
to emulate https://github.com/huggingface/tokenizers/blob/bb6fea033932ff1fb23807a2b7c5376f5a7c4786/tokenizers/src/tokenizer/mod.rs#L123-L125

mishig25 commented:
@Narsil please feel free to do another round of review. 
The things I've changed based on our huddle on friday are:
1. `process_chain` should not use/call `process`
2. `process_chain` should not be optional (i.e. every preprocessor must implement `process_chain` method)
3. added tests preprocessor (for bytelevel, template, sequence)

Narsil reviewed (commented):
Seems a fine direction but we need to remove the code within the `process` function in each implementor of the trait no ?

mishig25 commented on tokenizers/src/pre_tokenizers/byte_level.rs:
```diff
@@ -194,6 +194,24 @@ impl PostProcessor for ByteLevel {
+        }
+
+        Ok(encodings)
+    }
```
do you mean, we should change `process` to:
```rust
fn process(
    &self,
    mut encoding: Encoding,
    mut pair_encoding: Option<Encoding>,
    add_special_tokens: bool,
) -> Result<Encoding> {
    ...
    self.process_chain(...)
    ...
}
```
?

thomasw21 commented on bindings/python/src/processors.rs:
```diff
@@ -68,6 +70,14 @@ impl PostProcessor for PyPostProcessor {
             .process(encoding, pair_encoding, add_special_tokens)
     }
+
+    fn process_chain(
```
Ah so I see, thanks! Would be down to know if it would make sense to build such a feature where we can construct `private abstract methods`

SaulLu commented:
Hello to all! 

Thanks a lot again @mishig25 for taking the lead on this feature! :raised_hands: 

I was wondering where we are on this feature! What is currently the sticking point? Is there anything we need to discuss?

mishig25 commented:
Superceded by 1047
